kpops/pipeline_generator/pipeline.py

```diff
@@ -69,6 +80,47 @@ def __iter__(self) -> Iterator[PipelineComponent]:
     def __len__(self) -> int:
         return len(self.components)
 
+    def build_execution_graph_from(
+        self,
+        components: list[PipelineComponent],
+        reverse: bool,
+        runner: Callable[[PipelineComponent], Coroutine],
+    ):
+        async def run_graph_tasks(pending_tasks: list[Awaitable]):
+            for pending_task in pending_tasks:
+                await pending_task
+
+        nodes = [node_component.id for node_component in components]
+
+        transformed_graph = self.graph.subgraph(nodes)
+
+        if reverse:
+            transformed_graph = transformed_graph.reverse()
+
+        root_node = "root_node_bfs"
+        transformed_graph.add_node(root_node)
+
+        for node in self.graph:
+            predecessors = list(self.graph.predecessors(node))
+            if not predecessors:
+                transformed_graph.add_edge(root_node, node)
+
+        layers_graph = list(nx.bfs_layers(transformed_graph, root_node))
+
+        sorted_tasks = []
+
+        for layer in layers_graph[1:]:
```

Comment:
Why do yo start from index 1?

Reply:
because the first layer is the root node I added to transform the graph

Reply:
which is a fake node... it is just transforms the graph in other problem